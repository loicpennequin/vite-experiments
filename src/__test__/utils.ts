import { MockedFunction } from 'vitest';
import { render } from '@testing-library/vue';
import { App, Component, ref } from 'vue';
import { createRouter, createWebHistory } from 'vue-router';
import nock, { Body } from 'nock';
import defaultRoutes from '~pages';
import {
  IS_PRELOADING_INJECTION_KEY,
  LOADERS_INJECTION_KEY,
  POKEMON_API_URL
} from '../constants';
import { createQueryClient } from '../modules/app/factories/query-client.factory';
import { PluginModule } from '../types';

import { loaders } from '@/modules/app/plugins/loader.plugin';

export const renderWithPlugins = (
  component: Component,
  { routes = defaultRoutes, slots = {}, props = {} } = {}
) => {
  const router = createRouter({
    history: createWebHistory(),
    routes
  });
  const queryClient = createQueryClient();
  queryClient.mount();

  const pluginModules = import.meta.globEager<PluginModule>(
    '../**/*.plugin.ts'
  );
  Object.values(pluginModules)
    .map(p => p.default)
    .sort((a, b) => b.priority - a.priority)
    .forEach(plugin => {
      plugin.install?.({
        app: { use: () => ({}) } as unknown as App,
        router,
        isClient: true,
        initialState: undefined,
        initialRoute: undefined,
        meta: {
          queryClient
        }
      });
    });

  return render(component, {
    global: {
      plugins: [router],
      provide: {
        VUE_QUERY_CLIENT: queryClient,
        [LOADERS_INJECTION_KEY]: loaders,
        [IS_PRELOADING_INJECTION_KEY]: ref(false)
      }
    },
    slots,
    props
  });
};

type HttpMockOptions<T> = {
  url: string;
  statusCode?: number;
  response?: T;
  method?: 'get' | 'post' | 'put' | 'delete';
};

export const httpMock = <T extends Body>({
  url,
  statusCode = 200,
  response,
  method = 'get'
}: HttpMockOptions<T>) => {
  const mock = nock(POKEMON_API_URL).defaultReplyHeaders({
    'access-control-allow-origin': '*',
    'access-control-allow-credentials': 'true'
  });

  mock[method](url).reply(statusCode, response);

  return mock;
};

export const setDeviceWidth = (width: number) => {
  const defaultImplementation = (query: any) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn()
  });

  const mm = window.matchMedia as MockedFunction<any>;

  const getMatch = (query: string) => {
    const isWidthQuery = /(min-width|max-width)/.test(query);
    if (!isWidthQuery) return false;

    const number = query.match(/\d+/)?.[0];
    if (!number) return false;

    const isMinWidthQuery = /min-width/.test(query);

    return isMinWidthQuery ? width >= Number(number) : width <= Number(number);
  };

  mm.mockImplementation((query: any) => {
    return {
      matches: getMatch(query as string),
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn()
    };
  });

  afterEach(() => {
    mm.mockImplementation(defaultImplementation);
  });
};
