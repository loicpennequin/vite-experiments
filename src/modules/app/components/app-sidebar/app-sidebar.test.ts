import { httpMock, renderWithPlugins, setDeviceWidth } from '@/__test__/utils';
import Sidebar from './app-sidebar.vue';
import { waitFor, fireEvent } from '@testing-library/vue';

type PokemonEntry = {
  name: String;
};

type RenderWithLoadedListOptions = {
  results?: PokemonEntry[];
  isDesktop?: boolean;
  isOpened?: boolean;
};

async function renderWithLoadedList({
  results = [{ name: 'foo' }],
  isDesktop = false,
  isOpened = true
}: RenderWithLoadedListOptions = {}) {
  if (isDesktop) {
    setDeviceWidth(1920);
  }

  const request = httpMock({
    url: '/pokemon-species',
    query: true,
    response: {
      count: 905,
      results
    }
  });

  const wrapper = renderWithPlugins(Sidebar, { props: { isOpened } });
  await waitFor(() => expect(request.isDone()).toBeTruthy());

  return wrapper;
}

describe('Sidebar Component', () => {
  test('should mount component', async () => {
    expect(Sidebar).toBeTruthy();
    const { html } = await renderWithLoadedList({ isDesktop: true });

    expect(html()).toMatchSnapshot();
  });

  test('should display a link to detail page for each pokemon', async () => {
    const { findByText } = await renderWithLoadedList({
      isDesktop: true
    });

    const link = await findByText('foo');

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/foo');
  });

  test('should hide the list on mobile', async () => {
    const { queryByText } = await renderWithLoadedList({ isDesktop: false });
    expect(queryByText('foo')).not.toBeInTheDocument();
  });

  test('should display isOpened toggle on mobile', async () => {
    const { findByTitle } = await renderWithLoadedList({
      isDesktop: false
    });

    const toggle = await findByTitle('Hide list');
    expect(toggle).toBeInTheDocument();
  });
});
