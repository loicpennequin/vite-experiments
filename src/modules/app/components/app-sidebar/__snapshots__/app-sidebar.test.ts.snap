// Vitest Snapshot v1

exports[`Sidebar Component > should mount component 1`] = `
<nav p-y="3" bg="white dark:dark-100" min-w="15rem" flex="" flex-col="" shadow-md=""><label for="sidebar-toggle" sm="sr-only" title="Hide list" w="full" flex="" justify="start" items="center" gap="3" px="3"><svg width="1.2em" height="1.2em" viewBox="0 0 100 100" h="8">
      <use xlink:href="/assets/svg/svg-sprite-map.generated.svg#arrows-right"></use>
    </svg><span>Hide list</span></label>
  <div flex="1" overflow-y-auto="">
    <div class="infinite-scroll-trigger" data-v-dc2cc908=""></div>
  </div>
</nav>
`;
